README.md
**Shentry** is a single-file Python script which will run the wrapped
command and, if it fails, post an event to Sentry. By default, if the
wrapped script succeeds (exists with code 0), stdout/stderr are squashed,
similarly to [shuck](https://github.com/thwarted/shuck) or
[chronic](https://joeyh.name/code/moreutils/). It also always exits with
status 0 if events are able to be sent to Sentry.

It reads its configuration from the environment variable `$SHELL_SENTRY_DSN`
and, if such a variable is found, removes it from the environment before
calling the wrapped program. If that environment variable is not present, shentry will look
for `$SENTRY_DSN` (and similarly remove it from the environment).
If you need to use SENTRY_DSN inside your project code, make sure to set both.
You may also in that case want to put a top-level try/except around your whole
program to prevent uncaught exceptions from trigging both your in-process sentry sdk
and also your extra-process shentry, since you very likely only want one or the other.
If neither of the environment variables are present or both
are empty, shentry will try to read a DSN from `/etc/shentry_dsn`. If no DSN
can be found, the wrapped will have normal behavior (stdout/stderr will go
to their normal file descriptors, exit code will be passed through, etc).

This software should be compatible with Python 2.6, 2.7, and 3.5+;
that is to say, you should be able to run it just about anywhere.

[![Build Status](https://travis-ci.com/EasyPost/shentry.svg?branch=master)](https://travis-ci.com/EasyPost/shentry)


## Installation

Put the file [`shentry.py`](shentry.py) anywhere in your `$PATH` under the
name `shentry` and mark it as executable.

If the `requests` library is available, it will be used; otherwise, the standard
`urllib2` / `urllib.request` methods are used. If `requests` is used, then
either `$SHELL_SENTRY_PROXY` or the contents of `/etc/shentry_proxy` can be
used to configure an outbound proxy.

The `setup.py` in this directory only exists for `tox` to work (and run unit
tests). Don't bother using it; just copy `shentry.py` wherever you want it.

## Usage

You might want a crontab that looks something like the following:

    SHELL_SENTRY_DSN=https://pub:priv@app.getsentry.com/id

    15 * * * * /usr/local/bin/shentry /usr/local/bin/run-periodic-scripts

You can also make shentry your `$SHELL` and wrap all commands in it:

    SHELL_SENTRY_DSN=https://pub:priv@app.getsentry.com/id
    SHELL=/usr/local/bin/shentry

    15 * * * * /usr/local/bin/run-periodic-scripts
    7 1 * * * /usr/local/bin/run-daily-scripts

In this case, it will run the wrapped commands through `/bin/sh` (otherwise, it will honor `$SHELL`).

## License

This software is licensed under the ISC License, the full text of which can be found at [LICENSE.txt](LICENSE.txt).


CHANGES.md
0.4.1
-----
- Don't pop `$SENTRY_DSN` from the environment when we use it

0.4.0
-----
- pass through SIGTERM, SIGQUIT, and SIGINT to the child process
- do not send empty tags to Sentry
- (internal) switch tests from circleci to travisci

0.3.2
-----
- Move `level`, `server_name` and `sdk` from `tags` to top-level
- Add Python 3.6 to tox tests

0.3.1
-----
- Fix bug with loading large command outputs
- Add more tests

0.3.0
-----
- Add support for using `requests` if it's importable
- Add support for outbound proxies via `$SHELL_SENTRY_PROXY` or `/etc/shentry_proxy`


tests/test_basic.py
import os
import tempfile
import socket
import subprocess

import mock
import pytest

import shentry


class _Any(object):
    def __eq__(self, other):
        return True

    def __ne__(self, other):
        return False


ANY = _Any()


@pytest.mark.parametrize('argv,expected_full_command,expected_command_ws,expected_shell', (
    (['/foo/bar'], ['/bin/bash', '-c', '/foo/bar'], '/foo/bar', '/bin/bash'),
    (['/foo/bar', 'arg1', 'arg2'], ['/bin/bash', '-c', '/foo/bar arg1 arg2'], '/foo/bar arg1 arg2', '/bin/bash'),
    (['-c', 'ls | head'], ['/bin/sh', '-c', 'ls | head'], 'ls | head', '/bin/sh'),
))
def test_get_command(mocker, argv, expected_full_command, expected_command_ws, expected_shell):
    mocker.patch('os.environ', autospec=True)
    os.environ.get.return_value = '/bin/bash'
    assert shentry.get_command(argv) == (expected_full_command, expected_command_ws, expected_shell)
    os.environ.get.assert_called_once_with('SHELL', '/bin/sh')


class TestSimpleSentryClient(object):
    def test_new_from_environment(self, mocker):
        mocker.patch.dict('os.environ', {'SHELL_SENTRY_DSN': 'https://pub:priv@sentry.test/1'})
        client = shentry.SimpleSentryClient.new_from_environment()
        assert client.uri == 'https://sentry.test/api/1/store/'
        assert client.public == 'pub'
        assert client.secret == 'priv'
        assert client.project_id == '1'

    def test_new_from_environment_regular_dsn(self, mocker):
        mocker.patch.dict('os.environ', {'SENTRY_DSN': 'https://pub:priv@sentry.test/3'})
        client = shentry.SimpleSentryClient.new_from_environment()
        assert client.uri == 'https://sentry.test/api/3/store/'
        assert client.public == 'pub'
        assert client.secret == 'priv'
        assert client.project_id == '3'

    def test_new_from_environment_with_file(self, mocker):
        mocker.patch.dict('os.environ', {'SHELL_SENTRY_DSN': ''})
        mocker.patch.object(shentry, 'read_systemwide_config', return_value='https://pub:priv@sentry.test/2')
        client = shentry.SimpleSentryClient.new_from_environment()
        assert client.uri == 'https://sentry.test/api/2/store/'
        assert client.public == 'pub'
        assert client.secret == 'priv'
        assert client.project_id == '2'


def test_main(mocker, tmpdir):
    mock_client = mock.Mock(autospec=shentry.SimpleSentryClient)
    mocker.patch('shentry.SimpleSentryClient.new_from_environment', return_value=mock_client)
    mocker.patch('tempfile.mkdtemp', return_value=str(tmpdir))
    mocker.patch.dict('os.environ', {'SHELL': '/bin/fish', 'PATH': 'A_PATH', 'TZ': 'UTC'})
    mock_popen = mock.Mock(autospec=subprocess.Popen)
    mocker.patch('subprocess.Popen', return_value=mock_popen)
    mock_popen.wait.return_value = 1
    mock_popen.returncode = 1
    shentry.main(['shentry', '/bin/ls'])
    tempfile.mkdtemp.assert_called_once_with()
    mock_client.send_event.assert_called_once_with(
        message='Command `/bin/ls` failed with code 1.\n',
        level='error',
        fingerprint=[socket.gethostname(), '/bin/ls'],
        extra_context={
            'username': ANY,
            'shell': '/bin/fish',
            'load_average_at_exit': ANY,
            'start_time': ANY,
            'command': '/bin/ls',
            'duration': ANY,
            'PATH': 'A_PATH',
            'TZ': 'UTC',
            'returncode': 1,
            'working_directory': ANY,
            '_sent_with': ANY,
        }
    )


tests/test_integration.py
import collections
import threading
import subprocess
import os
import json
import mock
import pwd
import socket
import sys

import pytest

try:
    from BaseHTTPServer import HTTPServer
    from BaseHTTPServer import BaseHTTPRequestHandler
except ImportError:
    from http.server import HTTPServer
    from http.server import BaseHTTPRequestHandler


Request = collections.namedtuple('Request', ['command', 'path', 'headers', 'body'])


class SentryHTTPServer(HTTPServer):
    timeout = 0.1

    def __init__(self, *args, **kwargs):
        requests = kwargs.pop('requests')
        HTTPServer.__init__(self, *args, **kwargs)
        self.requests = requests

    def handle_timeout(self):
        pass


class SentryHTTPRequestHandler(BaseHTTPRequestHandler):
    def do_POST(self):
        body_len = int(self.headers.get('Content-Length', '0'))
        body = self.rfile.read(body_len)
        request = Request(command=self.command, path=self.path, headers=dict(self.headers.items()), body=body)
        self.server.requests.append(request)
        self.send_response(200)
        self.send_header('Content-Type', 'application/json')
        body = json.dumps({'status': 'ok'}).encode('utf-8')
        self.send_header('Content-Length', str(len(body)))
        self.end_headers()
        self.wfile.write(body)


class UUTHTTPServer(object):
    def __init__(self):
        self.running = False
        self.address = None
        self._thread = None
        self.requests = []
        self._started = threading.Condition()

    @property
    def uri(self):
        return 'http://sentry:password@{0}/'.format(':'.join(map(str, self.address)))

    def run(self):
        self.running = True
        httpd = SentryHTTPServer(('127.0.0.1', 0), SentryHTTPRequestHandler, requests=self.requests)
        self.address = httpd.server_address
        self._started.acquire()
        self._started.notify_all()
        self._started.release()
        while self.running:
            httpd.handle_request()

    def start(self):
        t = threading.Thread(target=self.run)
        self._thread = t
        t.start()
        self._started.acquire()
        self._started.wait()
        self._started.release()

    def stop(self):
        if self.running:
            self.running = False
            self._thread.join()


@pytest.yield_fixture
def http_server():
    t_s = UUTHTTPServer()
    t_s.start()
    yield t_s
    t_s.stop()


FAIL_NO_OUTPUT = '''#!/bin/bash

exit 1
'''

FAIL_LONG_OUTPUT = '''#!/bin/bash

for i in $(seq 1 4000)
do
    echo >&2 "line ${i}"
done

exit 1
'''


@pytest.fixture
def scripts(tmpdir):
    paths = {}
    for script in ('FAIL_NO_OUTPUT', 'FAIL_LONG_OUTPUT'):
        with open(os.path.join(str(tmpdir), script), 'w') as f:
            f.write(globals()[script])
            os.fchmod(f.fileno(), 0o700)
        paths[script] = os.path.join(str(tmpdir), script)
    return paths


def test_no_output(http_server, scripts):
    subprocess.check_call(
        [sys.executable, 'shentry.py', scripts['FAIL_NO_OUTPUT']],
        env={
            'SHELL_SENTRY_DSN': http_server.uri,
            'TZ': 'UTC',
        }
    )
    # ensure that the http server has processed all requests
    http_server.stop()
    assert len(http_server.requests) == 1
    req = http_server.requests[0]
    assert req.command == 'POST'
    body = json.loads(req.body.decode('utf-8'))
    assert body == {
        'device': mock.ANY,
        'event_id': mock.ANY,
        'extra': {
            'PATH': mock.ANY,
            'TZ': 'UTC',
            '_sent_with': mock.ANY,
            'command': scripts['FAIL_NO_OUTPUT'],
            'duration': mock.ANY,
            'load_average_at_exit': mock.ANY,
            'returncode': 1,
            'shell': '/bin/sh',
            'start_time': mock.ANY,
            'username': pwd.getpwuid(os.getuid()).pw_name,
            'working_directory': mock.ANY,
        },
        'fingerprint': mock.ANY,
        'message': 'Command `{0}` failed with code 1.\n'.format(scripts['FAIL_NO_OUTPUT']),
        'platform': 'other',
        'server_name': socket.gethostname(),
        'level': 'error',
        'sdk': {
            'name': 'shentry',
            'version': mock.ANY,
        },
        'timestamp': mock.ANY,
    }


def test_multi_kb_output(http_server, scripts):
    subprocess.check_call(
        [sys.executable, 'shentry.py', scripts['FAIL_LONG_OUTPUT']],
        env={
            'SHELL_SENTRY_DSN': http_server.uri,
            'TZ': 'UTC',
        }
    )
    # ensure that the http server has processed all requests
    http_server.stop()
    assert len(http_server.requests) == 1
    req = http_server.requests[0]
    assert req.command == 'POST'
    body = json.loads(req.body.decode('utf-8'))
    assert body == {
        'device': mock.ANY,
        'event_id': mock.ANY,
        'extra': {
            'PATH': mock.ANY,
            'TZ': 'UTC',
            '_sent_with': mock.ANY,
            'command': scripts['FAIL_LONG_OUTPUT'],
            'duration': mock.ANY,
            'load_average_at_exit': mock.ANY,
            'returncode': 1,
            'shell': '/bin/sh',
            'start_time': mock.ANY,
            'username': pwd.getpwuid(os.getuid()).pw_name,
            'working_directory': mock.ANY,
        },
        'fingerprint': mock.ANY,
        'message': mock.ANY,
        'platform': 'other',
        'server_name': socket.gethostname(),
        'level': 'error',
        'sdk': {
            'name': 'shentry',
            'version': mock.ANY,
        },
        'timestamp': mock.ANY,
    }
    expected = 'Command `{0}` failed with code 1.\n\nExcerpt of stderr:\n'.format(scripts['FAIL_LONG_OUTPUT'])
    assert body['message'].startswith(expected)


shentry.py
#!/usr/bin/env python

# Copyright (c) 2016, EasyPost <oss@easypost.com>
#
# Permission to use, copy, modify, and/or distribute this software for any purpose with or without fee is hereby
# granted, provided that the above copyright notice and this permission notice appear in all copies.
#
# THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES WITH REGARD TO THIS SOFTWARE INCLUDING ALL
# IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY SPECIAL, DIRECT,
# INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN
# AN ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR
# PERFORMANCE OF THIS SOFTWARE.


# NOTE: This code should work with Python 2.6, 2.7, 3.2, 3.3, 3.4, 3.5, and 3.6

from __future__ import print_function

try:
    from urllib.parse import urlparse
    from urllib.request import urlopen, Request
    from urllib.error import HTTPError, URLError
except ImportError:
    from urlparse import urlparse
    from urllib2 import urlopen, Request
    from urllib2 import HTTPError, URLError

try:
    import requests
    has_requests = True
except ImportError:
    has_requests = False


import datetime
import json
import os
import pwd
import sys
import time
import uuid
import subprocess
import tempfile
import shutil
import signal
import socket

from contextlib import closing

VERSION = '0.4.0'


def read_systemwide_config():
    try:
        with open('/etc/shentry_dsn', 'r') as f:
            return f.read().strip()
    except Exception:
        return None


def _get_proxy_url():
    if 'SHELL_SENTRY_PROXY' in os.environ:
        return os.environ['SHELL_SENTRY_PROXY']
    try:
        with open('/etc/shentry_proxy', 'r') as f:
            return f.read().strip()
    except Exception:
        pass
    return None


def _send_urllib2(uri, headers, data, timeout):
    req = Request(uri, data=data, headers=headers)
    try:
        with closing(urlopen(req, timeout=timeout)) as f:
            f.read()
        return True
    except HTTPError as e:
        print('Error {0} sending to Sentry'.format(e.code), file=sys.stderr)
        print(e.read(), file=sys.stderr)
        return False
    except URLError as e:
        print('Error {0} sending to Sentry'.format(e.reason), file=sys.stderr)
        return False


def _send_requests(uri, headers, data, timeout):
    try:
        kwargs = {}
        proxy_url = _get_proxy_url()
        if proxy_url is not None:
            kwargs['proxies'] = {
                'http': proxy_url,
                'https': proxy_url
            }
        resp = requests.post(
            uri, headers=headers, data=data, timeout=timeout,
            **kwargs
        )
        resp.raise_for_status()
        return True
    except requests.exceptions.RequestException as e:
        print('Error {0!r} sending to Sentry'.format(e), file=sys.stderr)
        return False


if has_requests:
    send_to_sentry = _send_requests
else:
    send_to_sentry = _send_urllib2


class SimpleSentryClient(object):
    TIMEOUT = 5
    SENTRY_VERSION = 5
    USER_AGENT = 'shentry/{0}'.format(VERSION)

    def __init__(self, dsn, uri, public, secret, project_id):
        self.dsn = dsn
        self.uri = uri
        self.public = public
        self.secret = secret
        self.project_id = project_id

    @classmethod
    def new_from_environment(cls):
        dsn = os.environ.pop('SHELL_SENTRY_DSN', '')
        if not dsn:
            dsn = os.environ.get('SENTRY_DSN', '')
        if not dsn:
            dsn = read_systemwide_config()
        if not dsn:
            return None
        else:
            try:
                dsn_fields = urlparse(dsn)
                keys, netloc = dsn_fields.netloc.split('@', 1)
                if ':' in keys:
                    public, private = keys.split(':', 1)
                else:
                    public = keys
                    private = ''
                project_id = dsn_fields.path.lstrip('/')
                uri = '{proto}://{netloc}/api/{project_id}/store/'.format(
                    proto=dsn_fields.scheme, netloc=netloc,
                    project_id=project_id,
                )
                return cls(dsn, uri, public, private, project_id)
            except Exception as e:
                print('Error parsing sentry DSN {0}: {1}'.format(dsn, e), file=sys.stderr)

    def send_event(self, message, level, fingerprint, logger='', culprit=None, extra_context={}):
        event_id = uuid.uuid4().hex
        now = int(time.time())
        uname = os.uname()
        event = {
            'event_id': event_id,
            'timestamp': datetime.datetime.utcnow().isoformat().split('.', 1)[0],
            'message': message,
            'level': level,
            'server_name': socket.gethostname(),
            'sdk': {
                'name': 'shentry',
                'version': VERSION,
            },
            'fingerprint': fingerprint,
            'platform': 'other',
            'device': {
                'name': uname[0],
                'version': uname[2],
                'build': uname[3]
            },
            'extra': {}
        }
        if logger:
            event['logger'] = logger
        if culprit is not None:
            event['culprit'] = culprit
        event['extra'].update(extra_context)
        headers = {
            'X-Sentry-Auth': (
                'Sentry sentry_version={self.SENTRY_VERSION}, '
                'sentry_client={self.USER_AGENT}, '
                'sentry_timestamp={now}, '
                'sentry_key={self.public}, '
                'sentry_secret={self.secret}'
            ).format(now=now, self=self),
            'User-Agent': self.USER_AGENT,
            'Content-Type': 'application/json',
        }
        if os.environ.get('SHELL_SENTRY_VERBOSE', '0') == '1':
            print('Sending to shentry', file=sys.stderr)
            print(event, file=sys.stderr)
        data = json.dumps(event).encode('utf-8')
        return send_to_sentry(uri=self.uri, headers=headers, data=data, timeout=self.TIMEOUT)


def get_command(argv):
    # get the command
    i_am_shell = False
    command = argv
    if command[0] == '-c':
        i_am_shell = True
        command = command[1:]
    if command[0] == '--':
        command = command[1:]
    shell = os.environ.get('SHELL', '/bin/sh')
    if i_am_shell or 'shentry' in shell:
        shell = '/bin/sh'
    command_ws = ' '.join(command)
    full_command = [shell, '-c', command_ws]
    return full_command, command_ws, shell


def show_usage():
    print('Usage: shentry [-c] command [...]', file=sys.stderr)
    print('', file=sys.stderr)
    print('Runs COMMAND, sending the output to Sentry if it exits non-0', file=sys.stderr)
    print('Takes sentry DSN from $SHELL_SENTRY_DSN, $SENTRY_DSN, or /etc/shentry_dsn', file=sys.stderr)


def read_snippet(fo, max_length):
    fo.seek(0, os.SEEK_END)
    length = fo.tell()
    rv = []
    fo.seek(0, os.SEEK_SET)
    read_all = False
    if length > max_length:
        top = int(max_length / 2) - 8
        bottom = max_length - top
        top = fo.read(top).decode('utf-8', 'ignore')
        rv.append(top)
        if not top.endswith('\n'):
            rv.append('\n')
        rv.append('\n[snip]\n')
        fo.seek(-1 * bottom, os.SEEK_END)
        rv.append(fo.read(bottom).decode('utf-8', 'ignore'))
    else:
        rv.append(fo.read().decode('utf-8', 'ignore'))
        read_all = True
    return ''.join(rv), read_all


def main(argv=None):
    if argv is None:
        argv = sys.argv
    if len(argv) < 2:
        show_usage()
        return 2
    extra_context = {
        'PATH': os.environ.get('PATH', ''),
        'username': pwd.getpwuid(os.getuid()).pw_name
    }
    if 'TZ' in os.environ:
        extra_context['TZ'] = os.environ['TZ']
    client = SimpleSentryClient.new_from_environment()
    full_command, command_ws, shell = get_command(argv[1:])
    extra_context['command'] = command_ws
    extra_context['shell'] = shell
    # if we couldn't configure sentry, just pass through
    if client is None:
        signal.signal(signal.SIGPIPE, signal.SIG_DFL)
        os.execv(shell, full_command)
        print('Unable to execv({0}, {1})'.format(shell, repr(full_command)), file=sys.stderr)
        return 1
    working_dir = None

    p = None

    def passthrough(signum, frame):
        if p is not None:
            p.send_signal(signum)
        else:
            raise ValueError('received signal %d without a child; bailing' % signum)

    def reset_signals():
        for sig in (signal.SIGTERM, signal.SIGQUIT, signal.SIGINT, signal.SIGPIPE):
            signal.signal(sig, signal.SIG_DFL)

    for sig in (signal.SIGTERM, signal.SIGQUIT, signal.SIGINT):
        signal.signal(sig, passthrough)
    try:
        working_dir = tempfile.mkdtemp()
        with open(os.path.join(working_dir, 'stdout'), 'w+b') as stdout:
            with open(os.path.join(working_dir, 'stderr'), 'w+b') as stderr:
                start_time = time.time()
                p = subprocess.Popen(full_command, stdout=stdout, stderr=stderr, shell=False,
                                     preexec_fn=reset_signals)
                extra_context['start_time'] = start_time
                extra_context['load_average_at_exit'] = ' '.join(map(str, os.getloadavg()))
                extra_context['working_directory'] = os.getcwd()
                extra_context['_sent_with'] = send_to_sentry.__name__
                if p.wait() != 0:
                    end_time = time.time()
                    extra_context['duration'] = end_time - start_time
                    code = p.returncode
                    extra_context['returncode'] = code
                    stderr_head, stderr_is_all = read_snippet(stderr, 700)
                    message = 'Command `{0}` failed with code {1}.\n'.format(command_ws, code)
                    if stderr_head:
                        if stderr_is_all:
                            message += '\nstderr:\n'
                        else:
                            message += '\nExcerpt of stderr:\n'
                        message += stderr_head
                    stdout_head, stdout_is_all = read_snippet(stdout, 200 + (700 - len(stderr_head)))
                    if stdout_head:
                        if stdout_is_all:
                            message += '\nstdout:\n'
                        else:
                            message += '\nExcerpt of stdout:\n'
                        message += stdout_head
                    client.send_event(
                        message=message,
                        level='error',
                        fingerprint=[socket.gethostname(), command_ws],
                        extra_context=extra_context,
                    )
    finally:
        if working_dir is not None:
            shutil.rmtree(working_dir)


if __name__ == '__main__':
    sys.exit(main())
